src/macaron/artifact/maven.py

```diff
@@ -0,0 +1,210 @@
+    This is because not all artifact types has a unique value of ``"classifier"`` according
+    to the Artifact Handlers table in the Maven Core reference. In addition, not supporting
+    the ``"classifier"`` qualifier at the moment simplifies the implementation for PURL
+    decoding and generation until there is a concrete use case for this additional qualifier.
+    """
+
+    # Enum with custom value type.
+    # See https://docs.python.org/3.10/library/enum.html#others.
+    JAR = _MavenArtifactType(
+        filename_pattern="{artifact_id}-{version}.jar",
+        purl_qualifiers={"type": "jar"},
+    )
+    POM = _MavenArtifactType(
+        filename_pattern="{artifact_id}-{version}.pom",
+        purl_qualifiers={"type": "pom"},
+    )
+    JAVADOC = _MavenArtifactType(
+        filename_pattern="{artifact_id}-{version}-javadoc.jar",
+        purl_qualifiers={"type": "javadoc"},
+    )
+    JAVA_SOURCE = _MavenArtifactType(
+        filename_pattern="{artifact_id}-{version}-sources.jar",
+        purl_qualifiers={"type": "java-source"},
+    )
+
+
+@dataclass
+class MavenArtifact:
+    """A Maven artifact."""
+
+    group_id: str
+    artifact_id: str
+    version: str
+    artifact_type: MavenArtifactType
+
+    @property
+    def package_url(self) -> PackageURL:
+        """Get the PackageURL of this Maven artifact."""
+        return PackageURL(
+            type="maven",
+            namespace=self.group_id,
+            name=self.artifact_id,
+            version=self.version,
+            qualifiers=self.artifact_type.purl_qualifiers,
+        )
+
+    @classmethod
+    def from_package_url(cls, package_url: PackageURL) -> Self | None:
+        """Create a Maven artifact from a PackageURL.
+
+        Parameters
+        ----------
+        package_url : PackageURL
+            The PackageURL identifying a Maven artifact.
+
+        Returns
+        -------
+        Self | None
+            A Maven artifact, or ``None`` if the PURL is not a valid Maven artifact PURL, or if
+            the artifact type is not supported.
+            For supported artifact types, see :class:`MavenArtifactType`.
+        """
+        if not package_url.namespace:
+            return None
+        if not package_url.version:
+            return None
+        if package_url.type != "maven":
+            return None
+        maven_artifact_type = None
+        for artifact_type in MavenArtifactType:
+            if artifact_type.purl_qualifiers == package_url.qualifiers:
+                maven_artifact_type = artifact_type
+                break
+        if not maven_artifact_type:
+            return None
+        return cls(
+            group_id=package_url.namespace,
+            artifact_id=package_url.name,
+            version=package_url.version,
+            artifact_type=maven_artifact_type,
+        )
+
+    @classmethod
+    def from_artifact_name(
+        cls,
+        artifact_name: str,
```

Comment:
`artifact_name` seems a bit ambiguous. Is it the filename?

Reply:
It is indeed the filename. Thanks for pointing out.
I have renamed `artifact_name` to `artifact_filename` in 263414bbd7cdd94a9613adcc97a66453d988a22f.